### ISSUE TITLE
TEST-46-HOMED test failure

### ISSUE THREAD

**Issue by mbiebl** [systemd/systemd#31518]
Version: latest git master

Running the upstream autopkgtest on Debian sid yields

```
[  271.421790] testsuite-46.sh[58]: + systemctl daemon-reload
[  271.511616] (sd-exec-[688]: /usr/lib/systemd/system-generators/systemd-ssh-generator failed with exit status 1.
[  271.574476] testsuite-46.sh[58]: + systemctl start mysshserver.socket
[  271.584057] systemd[1]: Cannot find unit for notify message of PID 681, ignoring.
[  271.590216] testsuite-46.sh[58]: + userdbctl user -j homedsshtest
[  271.592365] systemd[1]: Cannot find unit for notify message of PID 704, ignoring.
[  271.594474] testsuite-46.sh[705]: {
[  271.594474] testsuite-46.sh[705]:    "luksDiscard" : true,
[  271.594474] testsuite-46.sh[705]:    "luksPbkdfType" : "pbkdf2",
[  271.594474] testsuite-46.sh[705]:    "luksPbkdfTimeCostUSec" : 1000,
[  271.594474] testsuite-46.sh[705]:    "enforcePasswordPolicy" : false,
[  271.594474] testsuite-46.sh[705]:    "stopDelayUSec" : 0,
[  271.594474] testsuite-46.sh[705]:    "userName" : "homedsshtest",
[  271.594474] testsuite-46.sh[705]:    "perMachine" : [
[  271.594474] testsuite-46.sh[705]:            {
[  271.594474] testsuite-46.sh[705]:                    "diskSize" : 0,
[  271.594474] testsuite-46.sh[705]:                    "rebalanceWeight" : 0,
[  271.594474] testsuite-46.sh[705]:                    "matchMachineId" : "120516797c5647c8928aae160230ff10"
[  271.594474] testsuite-46.sh[705]:            }
[  271.594474] testsuite-46.sh[705]:    ],
[  271.594474] testsuite-46.sh[705]:    "disposition" : "regular",
[  271.594474] testsuite-46.sh[705]:    "lastChangeUSec" : 1709080837298801,
[  271.594474] testsuite-46.sh[705]:    "lastPasswordChangeUSec" : 1709080837298801,
[  271.594474] testsuite-46.sh[705]:    "privileged" : {
[  271.594474] testsuite-46.sh[705]:            "sshAuthorizedKeys" : [
[  271.594474] testsuite-46.sh[705]:                    "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQC9UNZHN7Pc8juj6PW/oIfjbvDoEo6emRP3SeU2IusovXD+Hy5WgP4fHPntyoHz9y0R5NQHhViv5UjCyoRoB90z59UahZ2v7XrGC+eAnuJE4+OUlwOAvQ2Sdn0k3S1N+v97yNviFPXAUzrEE+dS+g+GrTQoXmjMTKBeWig6V83rCyMW4aCNagF+GB4MQdcnKQYNGR2iZglKtMPfNV1Z0/dSYEJasn1pinCOBkH0EiZYzm4ihjSxAuHdEeRmgoyqhHgyXPhmjXZKF64ZfkYUxqiUL09cDBU/J/LrTcHEhgCCExOa3x7kv91D1PLA5BSnZEIqDSm23aVetHTdNoYIWkdKIk9uZqz8v6pNQDXg/hv1yMjFFbptGnFV+6JStFYTqX1S16yZpK8u4GLTFVI1q3mwQ32LzHL5To3zOvPmRiN0u2n7NdcuJJP8Wyifc7MGS/sZTGT1u7c2sCHexf7lC3d4l1wCBvjL68SiTUD3VBMR7uPmiV8AE6TDb4p0fc857Gs= "
[  271.594474] testsuite-46.sh[705]:            ],
[  271.594474] testsuite-46.sh[705]:            "hashedPassword" : [
[  271.594474] testsuite-46.sh[705]:                    "$y$j9T$lGqvHdphJIZ2yTSeQudfv.$eZICZU.k.o.9B00R43ZT5OtHH6QzQd.DVPt0.56eIq5"
[  271.594474] testsuite-46.sh[705]:            ]
[  271.594474] testsuite-46.sh[705]:    },
[  271.594474] testsuite-46.sh[705]:    "signature" : [
[  271.594474] testsuite-46.sh[705]:            {
[  271.594474] testsuite-46.sh[705]:                    "data" : "xenP/HLlpNSGxs3Y/V2nfr8RruuQeuNYH+GKseARyqhHPGoRYqI4rgiMbUY3hABML2SWyasszgJimtLaZvpQBg==",
[  271.594474] testsuite-46.sh[705]:                    "key" : "-----BEGIN PUBLIC KEY-----\nMCowBQYDK2VwAyEAg0WRu9k6L75hIi3ZqBzJjymds68BZjLkiruro+UKxzU=\n-----END PUBLIC KEY-----\n"
[  271.594474] testsuite-46.sh[705]:            }
[  271.594474] testsuite-46.sh[705]:    ],
[  271.594474] testsuite-46.sh[705]:    "binding" : {
[  271.594474] testsuite-46.sh[705]:            "120516797c5647c8928aae160230ff10" : {
[  271.594474] testsuite-46.sh[705]:                    "blobDirectory" : "/var/cache/systemd/home/homedsshtest",
[  271.594474] testsuite-46.sh[705]:                    "imagePath" : "/home/homedsshtest.homedir",
[  271.594474] testsuite-46.sh[705]:                    "uid" : 60202,
[  271.594474] testsuite-46.sh[705]:                    "gid" : 60202,
[  271.594474] testsuite-46.sh[705]:                    "storage" : "directory"
[  271.594474] testsuite-46.sh[705]:            }
[  271.594474] testsuite-46.sh[705]:    },
[  271.594474] testsuite-46.sh[705]:    "status" : {
[  271.594474] testsuite-46.sh[705]:            "120516797c5647c8928aae160230ff10" : {
[  271.594474] testsuite-46.sh[705]:                    "state" : "inactive",
[  271.594474] testsuite-46.sh[705]:                    "service" : "io.systemd.Home",
[  271.594474] testsuite-46.sh[705]:                    "useFallback" : true,
[  271.594474] testsuite-46.sh[705]:                    "fallbackShell" : "/usr/bin/systemd-home-fallback-shell",
[  271.594474] testsuite-46.sh[705]:                    "fallbackHomeDirectory" : "/",
[  271.594474] testsuite-46.sh[705]:                    "diskSize" : 304087040,
[  271.594474] testsuite-46.sh[705]:                    "diskFree" : 304082944,
[  271.594474] testsuite-46.sh[705]:                    "diskCeiling" : 5497558138880,
[  271.594474] testsuite-46.sh[705]:                    "diskFloor" : 5242880,
[  271.594474] testsuite-46.sh[705]:                    "signedLocally" : true,
[  271.594474] testsuite-46.sh[705]:                    "fileSystemType" : "tmpfs"
[  271.594474] testsuite-46.sh[705]:            }
[  271.594474] testsuite-46.sh[705]:    }
[  271.594474] testsuite-46.sh[705]: }
[  271.595790] testsuite-46.sh[706]: + ssh -t -t -4 -p 4711 -i /tmp/homed.id_rsa -o 'SetEnv PASSWORD=hunter4711' -o 'StrictHostKeyChecking no' homedsshtest@localhost echo zzz
[  271.596126] testsuite-46.sh[708]: + tr -d '\r'
[  271.596323] testsuite-46.sh[707]: + tail -n 1
[  271.630737] testsuite-46.sh[706]: kex_exchange_identification: read: Connection reset by peer
[  271.630797] testsuite-46.sh[706]: Connection reset by 127.0.0.1 port 4711
[  271.632227] testsuite-46.sh[58]: + at_exit
[  271.632227] testsuite-46.sh[58]: + systemctl stop mysshserver.socket
[  271.642498] testsuite-46.sh[58]: + rm -f /tmp/homed.id_rsa /run/systemd/system/mysshserver.socket /run/systemd/system/mysshserver@.service
[  271.643982] systemd[1]: Cannot find unit for notify message of PID 712, ignoring.
[  271.644294] testsuite-46.sh[58]: + systemctl daemon-reload
[  271.727277] (sd-exec-[721]: /usr/lib/systemd/system-generators/systemd-ssh-generator failed with exit status 1.
[  271.798209] testsuite-46.sh[58]: + homectl remove homedsshtest
[  271.804106] systemd[1]: Cannot find unit for notify message of PID 714, ignoring.
[  271.810144] testsuite-46.sh[58]: + mv /etc/pam.d/sshd.save46 /etc/pam.d/sshd
[  271.812474] systemd[1]: testsuite-46.service: Failed with result 'exit-code'.
[  271.812846] systemd[1]: Failed to start testsuite-46.service.
Finishing after writing 22034 entries
TEST-46-HOMED: (failed; see logs)
-rw-r----- 1 root root 16777216 Feb 28 00:40 /tmp/autopkgtest.8I5fod/upstream-artifacts/TEST-46-HOMED.journal
nspawn-root test failed
make: Leaving directory '/tmp/autopkgtest.8I5fod/tree/test/TEST-46-HOMED'
[00:40:39] --x-- Result of TEST-46-HOMED: 2 --x--
make: *** [Makefile:4: run] Error 1

TEST-46-HOMED:                      FAIL     ( 42 s)
```
Full test log attached
[log.txt](https://github.com/systemd/systemd/files/14427213/log.txt)

**Comment by yuwata** [comment]
Similar failure on CentOS CI:
https://jenkins-systemd.apps.ocp.cloud.ci.centos.org/job/upstream-vagrant-archlinux/11245/artifact/systemd-centos-ci/index.html

**Comment by AdrianVovk** [comment]
Why reopen?

**Comment by bluca** [comment]
Still not fixed, see https://github.com/systemd/systemd/pull/31544

**Comment by bluca** [comment]
lol, lmao: https://salsa.debian.org/systemd-team/systemd/-/blob/debian/master/debian/extra/checkout-upstream?ref_type=heads#L103

**Comment by mbiebl** [comment]
fwiw, I did not use checkout-upstream to build the debs.

**Comment by mbiebl** [comment]
I already noticed 
```
dh_missing: warning: etc/ssh/sshd_config.d/20-systemd-userdb.conf exists in debian/tmp but is not installed to anywhere 
dh_missing: warning: usr/lib/systemd/sshd_config.d/20-systemd-userdb.conf exists in debian/tmp but is not installed to anywhere 
dh_missing: warning: usr/lib/tmpfiles.d/20-systemd-userdb.conf exists in debian/tmp but is not installed to anywhere 
```
and included them in my local build.
The test still fails :-/

**Comment by bluca** [comment]
Was that main built in sid? With the upstream profile?

The test works in noble/jammy with those changes: https://github.com/systemd/systemd/pull/31527

**Comment by mbiebl** [comment]
yep, with Debian sid and current HEAD, i.e. 0abe0cd402

**Comment by mbiebl** [comment]
[journal.txt](https://github.com/systemd/systemd/files/14460356/journal.txt)
[log.txt](https://github.com/systemd/systemd/files/14460357/log.txt)

**Comment by mrc0mmand** [comment]
Hmpf, this is strange, since I can't reproduce the issue on Debian Sid after building systemd directly from the upstream repo.

**Comment by mbiebl** [comment]
@mrc0mmand: good news is: after reverting #31403 it finally passes for me.

**Comment by mrc0mmand** [comment]
@mbiebl just out of curiosity, what's the output of `dpkg-query -l | grep libpam0` on your Sid machine?

I wanted to play with this a bit more in a fresh VM, but now I can't even install build dependencies on bookworm:

```
# apt build-dep -y systemd
Reading package lists... Done
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 builddeps:systemd : Depends: tzdata-legacy but it is not installable or
                              tzdata (< 2023c-8) but 2024a-0+deb12u1 is to be installed
E: Unable to correct problems, you have held broken packages.
```

And Sid seems to be caught up in the middle of `libpam0g` -> `libpam0t64` transition, so my system just ends up in some weird in-between state without working PAM:

```
# apt install libpam0g-dev
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 util-linux : PreDepends: libpam0t64 (>= 0.99.7.1) but it is not installable
E: Error, pkgProblemResolver::Resolve generated breaks, this may be caused by held packages.
```
(there doesn't seem to be `libpam0t64-dev` package)

**Comment by bluca** [comment]
ignore the tzdata thing, it was fixed later in git, installing tzdata is enough - and yeah avoid sid like the plague for the next couple of weeks, everything is literally on fire

**Comment by mbiebl** [comment]
@mrc0mmand The time_t-64 transition started this weekend and you are caught in the middle of it. As @bluca said, unstable deserves its name right now.
I haven't updated my system and test VMs from my last weeks state delibertately. So I can run (build-)tests and stuff on sid. I don't have any t64 packages installed atm.

**Comment by mbiebl** [comment]
> Hmpf, this is strange, since I can't reproduce the issue on Debian Sid after building systemd directly from the upstream repo.

how do you run Debian? chroot, container, qemu?

**Comment by mrc0mmand** [comment]
> > Hmpf, this is strange, since I can't reproduce the issue on Debian Sid after building systemd directly from the upstream repo.
> 
> how do you run Debian? chroot, container, qemu?

It's a QEMU (KVM) VM.

Anyway, out of curiosity I spun up a Ubuntu Noble VM and I can reproduce the issue there if I run the test several times in a succession. Which also reproduces the issue on my Fedora machine, after a couple dozen retries, so it looks like a race.

**Comment by mrc0mmand** [comment]
And what's more interesting is that I can now reproduce the issue even with https://github.com/systemd/systemd/pull/31403 reverted, so that's, unfortunately, not the root cause.

Also, with #31403 reverted the PAM errors it fixes are now back, but interestingly enough they show up only in the failing case.

Successful run:
```
[ 3836.475962] H testsuite-46.sh[703]: + ssh -t -t -4 -p 4711 -i /tmp/homed.id_ecdsa -o 'SetEnv PASSWORD=hunter4711' -o 'StrictHostKeyChecking no' homedsshtest@localhost echo zzz
[ 3836.476184] H testsuite-46.sh[704]: + tr -d '\r'
[ 3836.476546] H testsuite-46.sh[705]: + tee /tmp/homedsshtest.out
[ 3836.490251] H systemd[1]: Created slice system-mysshserver.slice.
[ 3836.491858] H systemd[1]: Started mysshserver@0-127.0.0.1:4711-127.0.0.1:42840.service.
...
[ 3836.741481] H systemd-homed[41]: homedsshtest: changing state inactive → activating-for-acquire
[ 3836.751007] H systemd-homework[710]: None of the supplied plaintext passwords unlock the user record's hashed passwords.
[ 3836.769831] H systemd-homed[41]: Activation failed: Password for home homedsshtest is incorrect or not sufficient for authentication.
[ 3836.770041] H systemd-homed[41]: homedsshtest: changing state activating-for-acquire → inactive
[ 3836.770128] H systemd-homed[41]: Got notification that all sessions of user homedsshtest ended, deactivating automatically.
[ 3836.770189] H systemd-homed[41]: Home homedsshtest already deactivated, no automatic deactivation needed.
[ 3836.770788] H sshd[706]: pam_systemd_home(sshd:account): Home for user homedsshtest acquired in incomplete mode, requires later activation.
[ 3836.771052] H sshd[706]: Accepted publickey for homedsshtest from 127.0.0.1 port 42840 ssh2: ECDSA SHA256:rhydFeN3skqryLJk/d4eRmbLYTfFM8+l+NIDjIiLDHU
...
[ 3836.782986] H systemd[1]: Created slice user-60202.slice.
[ 3836.784682] H systemd[1]: Starting user-runtime-dir@60202.service...
[ 3836.787511] H systemd-logind[42]: New session c1 of user homedsshtest.
[ 3836.835244] H systemd[1]: Finished user-runtime-dir@60202.service.
[ 3836.873162] H systemd[1]: Started session-c1.scope.
[ 3836.875886] H sshd[706]: pam_unix(sshd:session): session opened for user homedsshtest(uid=60202) by (uid=0)
[ 3836.876786] H sshd[706]: User child is on pid 713
...
[ 3836.974050] H systemd-homed[41]: homedsshtest: changing state inactive → activating
[ 3837.016728] H systemd-homework[715]: Provided password unlocks user record.
[ 3837.017839] H systemd-homework[715]: Read embedded .identity file.
[ 3837.040097] H systemd-homework[715]: Provided password unlocks user record.
[ 3837.040317] H systemd-homework[715]: Reconciling embedded user identity completed (host and embedded version were identical).
[ 3837.041721] H systemd-homework[715]: Recursive changing of ownership not necessary, skipped.
[ 3837.041799] H systemd-homework[715]: Synchronized disk.
[ 3837.042272] H systemd-homework[715]: Moving to final mount point /home/homedsshtest completed.
[ 3837.042379] H systemd-homework[715]: Everything completed.
[ 3837.044701] H systemd-homed[41]: Home homedsshtest is signed exclusively by our key, accepting.
[ 3837.063451] H systemd-homed[41]: homedsshtest: changing state activating → active
[ 3837.103011] H systemd[1]: Starting user@60202.service...
[ 3837.128528] H (systemd)[717]: pam_systemd_home(systemd-user:account): Home for user homedsshtest successfully acquired.
[ 3837.149725] H systemd-logind[42]: New session 157 of user homedsshtest.
[ 3837.215105] H systemd[717]: Queued start job for default target default.target.
...
[ 3837.285751] H systemd[717]: Startup finished in 128ms.
...
[ 3837.289110] H sshd[713]: debug1: Received SIGCHLD.
[ 3837.289456] H testsuite-46.sh[703]: Connection to localhost closed.
...
[ 3837.292713] H sshd[706]: debug1: do_cleanup
[ 3837.292713] H sshd[706]: debug1: PAM: cleanup
[ 3837.292713] H sshd[706]: debug1: PAM: closing session
[ 3837.293012] H sshd[706]: pam_systemd_home(sshd:session): Not deactivating home directory of homedsshtest, as it is still used.
[ 3837.293919] H sshd[706]: pam_unix(sshd:session): session closed for user homedsshtest
[ 3837.294010] H sshd[706]: debug1: PAM: deleting credentials
...
[ 3837.297966] H testsuite-46.sh[52]: + ssh -t -t -4 -p 4711 -i /tmp/homed.id_ecdsa -o 'SetEnv PASSWORD=hunter4711' -o 'StrictHostKeyChecking no' homedsshtest@localhost env
[ 3837.298052] H sshd[706]: debug1: restore_uid: 0/0
[ 3837.298052] H sshd[706]: debug1: audit_event: unhandled event 12
[ 3837.299568] H systemd[1]: mysshserver@0-127.0.0.1:4711-127.0.0.1:42840.service: Deactivated successfully.
[ 3837.303378] H systemd[1]: session-c1.scope: Deactivated successfully.
[ 3837.312706] H systemd[1]: Started mysshserver@1-127.0.0.1:4711-127.0.0.1:42850.service.
...
[ 3837.338960] H systemd[1]: Stopping user@60202.service...
[ 3837.341995] H systemd[717]: Activating special unit exit.target...
[ 3837.342632] H systemd[717]: Stopped target default.target.
...
[ 3837.370009] H systemd-homed[41]: Got notification that all sessions of user homedsshtest ended, deactivating automatically.
[ 3837.371736] H systemd-homed[41]: homedsshtest: changing state active → deactivating
[ 3837.380553] H systemd[1]: user@60202.service: Deactivated successfully.
[ 3837.380972] H systemd[1]: Stopped user@60202.service.
[ 3837.382623] H systemd-homework[733]: Syncing completed.
[ 3837.382754] H systemd-homework[733]: Unmounting completed.
[ 3837.382908] H systemd-homework[733]: Everything completed.
[ 3837.383169] H systemd-homed[41]: homedsshtest: changing state deactivating → inactive
...
[ 3837.554400] H systemd[1]: Removed slice user-60202.slice.
[ 3837.564737] H sshd[730]: debug1: do_pam_account: called
[ 3837.589586] H systemd-homed[41]: homedsshtest: changing state inactive → activating-for-acquire
[ 3837.598823] H systemd-homework[740]: None of the supplied plaintext passwords unlock the user record's hashed passwords.
[ 3837.622429] H systemd-homed[41]: Activation failed: Password for home homedsshtest is incorrect or not sufficient for authentication.
[ 3837.623042] H systemd-homed[41]: homedsshtest: changing state activating-for-acquire → inactive
[ 3837.623223] H systemd-homed[41]: Got notification that all sessions of user homedsshtest ended, deactivating automatically.
[ 3837.623296] H systemd-homed[41]: Home homedsshtest already deactivated, no automatic deactivation needed.
[ 3837.624058] H sshd[730]: pam_systemd_home(sshd:account): Home for user homedsshtest acquired in incomplete mode, requires later activation.
[ 3837.624265] H sshd[730]: Accepted publickey for homedsshtest from 127.0.0.1 port 42850 ssh2: ECDSA SHA256:rhydFeN3skqryLJk/d4eRmbLYTfFM8+l+NIDjIiLDHU
...
[ 3837.639242] H systemd[1]: Created slice user-60202.slice.
[ 3837.641967] H systemd[1]: Starting user-runtime-dir@60202.service...
[ 3837.674167] H systemd-logind[42]: New session c2 of user homedsshtest.
[ 3837.682789] H systemd[1]: Finished user-runtime-dir@60202.service.
[ 3837.686786] H systemd[1]: Started session-c2.scope.
[ 3837.689676] H sshd[730]: pam_unix(sshd:session): session opened for user homedsshtest(uid=60202) by (uid=0)
...
[ 3837.775919] H systemd-homed[41]: homedsshtest: changing state inactive → activating
[ 3837.823182] H systemd-homework[745]: Provided password unlocks user record.
[ 3837.824213] H systemd-homework[745]: Read embedded .identity file.
[ 3837.845508] H systemd-homework[745]: Provided password unlocks user record.
[ 3837.845651] H systemd-homework[745]: Reconciling embedded user identity completed (host and embedded version were identical).
[ 3837.846977] H systemd-homework[745]: Recursive changing of ownership not necessary, skipped.
[ 3837.847014] H systemd-homework[745]: Synchronized disk.
[ 3837.847593] H systemd-homework[745]: Moving to final mount point /home/homedsshtest completed.
[ 3837.847637] H systemd-homework[745]: Everything completed.
[ 3837.849787] H systemd-homed[41]: Home homedsshtest is signed exclusively by our key, accepting.
[ 3837.873553] H systemd-homed[41]: homedsshtest: changing state activating → active
[ 3837.906866] H systemd[1]: Starting user@60202.service...
[ 3837.940937] H (systemd)[747]: pam_systemd_home(systemd-user:account): Home for user homedsshtest successfully acquired.
[ 3837.960380] H systemd-logind[42]: New session 158 of user homedsshtest.
[ 3838.052974] H systemd[747]: Queued start job for default target default.target.
...
[ 3838.100661] H systemd[1]: Started user@60202.service.
...
[ 3838.106919] H testsuite-46.sh[727]: Connection to localhost closed.
```

Failed run:
```
[ 3854.123323] H testsuite-46.sh[702]: + ssh -t -t -4 -p 4711 -i /tmp/homed.id_ecdsa -o 'SetEnv PASSWORD=hunter4711' -o 'StrictHostKeyChecking no' homedsshtest@localhost echo zzz
[ 3854.123627] H testsuite-46.sh[703]: + tr -d '\r'
[ 3854.123889] H testsuite-46.sh[704]: + tee /tmp/homedsshtest.out
[ 3854.132087] H systemd[1]: Created slice system-mysshserver.slice.
[ 3854.133857] H systemd[1]: Started mysshserver@0-127.0.0.1:4711-127.0.0.1:45548.service.
...
[ 3854.363161] H sshd[705]: debug1: do_pam_account: called
[ 3854.394593] H systemd-homed[41]: homedsshtest: changing state inactive → activating-for-acquire
[ 3854.403284] H systemd-homework[709]: None of the supplied plaintext passwords unlock the user record's hashed passwords.
[ 3854.429615] H systemd-homed[41]: Activation failed: Password for home homedsshtest is incorrect or not sufficient for authentication.
[ 3854.429951] H systemd-homed[41]: homedsshtest: changing state activating-for-acquire → inactive
[ 3854.430021] H systemd-homed[41]: Got notification that all sessions of user homedsshtest ended, deactivating automatically.
[ 3854.430117] H systemd-homed[41]: Home homedsshtest already deactivated, no automatic deactivation needed.
[ 3854.431110] H sshd[705]: pam_systemd_home(sshd:account): Home for user homedsshtest acquired in incomplete mode, requires later activation.
[ 3854.431557] H sshd[705]: Accepted publickey for homedsshtest from 127.0.0.1 port 45548 ssh2: ECDSA SHA256:wPPqPc0e3VA+q8dAXFR8i36yDlXGlvtfrZF23hAH2Hw
...
[ 3854.433954] H sshd[705]: debug1: PAM: establishing credentials
[ 3854.454498] H systemd-logind[42]: New session c1 of user homedsshtest.
[ 3854.456351] H systemd[1]: Created slice user-60202.slice.
[ 3854.459113] H systemd[1]: Starting user-runtime-dir@60202.service...
[ 3854.504329] H systemd[1]: Finished user-runtime-dir@60202.service.
[ 3854.507791] H systemd[1]: Started session-c1.scope.
[ 3854.510154] H sshd[705]: pam_unix(sshd:session): session opened for user homedsshtest(uid=60202) by (uid=0)
[ 3854.510649] H sshd[705]: User child is on pid 712
...
[ 3854.612109] H systemd-homed[41]: homedsshtest: changing state inactive → activating
[ 3854.637129] H systemd-homework[714]: Provided password unlocks user record.
[ 3854.638194] H systemd-homework[714]: Read embedded .identity file.
[ 3854.657622] H systemd-homework[714]: Provided password unlocks user record.
[ 3854.657795] H systemd-homework[714]: Reconciling embedded user identity completed (host and embedded version were identical).
[ 3854.668919] H systemd-homework[714]: Recursive changing of ownership not necessary, skipped.
[ 3854.668972] H systemd-homework[714]: Synchronized disk.
[ 3854.673877] H systemd-homework[714]: Moving to final mount point /home/homedsshtest completed.
[ 3854.673941] H systemd-homework[714]: Everything completed.
[ 3854.676563] H systemd-homed[41]: Home homedsshtest is signed exclusively by our key, accepting.
[ 3854.709906] H systemd-homed[41]: homedsshtest: changing state activating → active
[ 3854.737933] H systemd[1]: Starting user@60202.service...
[ 3854.769186] H (systemd)[716]: pam_systemd_home(systemd-user:account): Home for user homedsshtest successfully acquired.
[ 3854.787307] H systemd-logind[42]: New session 159 of user homedsshtest.
[ 3854.868267] H systemd[716]: Queued start job for default target default.target.
...
[ 3854.933235] H systemd[716]: Startup finished in 137ms.
[ 3854.933566] H systemd[1]: Started user@60202.service.
...
[ 3854.937803] H testsuite-46.sh[702]: Connection to localhost closed.
...
[ 3854.941317] H sshd[705]: debug1: do_cleanup
[ 3854.941317] H sshd[705]: debug1: PAM: cleanup
[ 3854.941317] H sshd[705]: debug1: PAM: closing session
[ 3854.941317] H sshd[705]: debug1: PAM: deleting credentials
[ 3854.941791] H sshd[705]: pam_systemd_home(sshd:session): Not deactivating home directory of homedsshtest, as it is still used.
[ 3854.942106] H sshd[705]: debug1: temporarily_use_uid: 60202/60202 (e=0/0)
[ 3854.942785] H testsuite-46.sh[52]: + rm /tmp/homedsshtest.out
[ 3854.942910] H sshd[705]: pam_unix(sshd:session): session closed for user homedsshtest
[ 3854.944264] H testsuite-46.sh[52]: + ssh -t -t -4 -p 4711 -i /tmp/homed.id_ecdsa -o 'SetEnv PASSWORD=hunter4711' -o 'StrictHostKeyChecking no' homedsshtest@localhost env
[ 3854.946868] H systemd[1]: mysshserver@0-127.0.0.1:4711-127.0.0.1:45548.service: Deactivated successfully.
[ 3854.948547] H systemd[1]: session-c1.scope: Deactivated successfully.
[ 3854.950759] H systemd-logind[42]: Removed session c1.
[ 3854.953756] H systemd[1]: Stopping user@60202.service...
[ 3854.955036] H systemd[716]: Activating special unit exit.target...
...
[ 3854.958981] H systemd[716]: Reached target exit.target.
[ 3854.960557] H systemd[1]: Started mysshserver@1-127.0.0.1:4711-127.0.0.1:45562.service.
[ 3854.981192] H (sd-pam)[717]: pam_systemd_home(systemd-user:session): Failed to release user home: Access denied
[ 3854.981510] H dbus-daemon[45]: [system] Rejected send message, 2 matched rules; type="method_call", sender=":1.87" (uid=60202 pid=717 comm="(sd-pam)" label="unconfined") interface="org.freedesktop.login1.Manager" member="ReleaseSession>
[ 3854.981896] H (sd-pam)[717]: pam_systemd(systemd-user:session): Failed to release session: Access denied
[ 3854.983156] H systemd-homed[41]: Got notification that all sessions of user homedsshtest ended, deactivating automatically.
[ 3854.983559] H systemd-homed[41]: homedsshtest: changing state active → deactivating
[ 3854.986393] H systemd[1]: user@60202.service: Deactivated successfully.
[ 3854.986785] H systemd[1]: Stopped user@60202.service.
[ 3854.988222] H systemd-homework[731]: Syncing completed.
[ 3854.988336] H systemd-homework[731]: Unmounting completed.
[ 3854.988416] H systemd-homework[731]: Everything completed.
[ 3854.988872] H systemd[1]: home-homedsshtest.mount: Deactivated successfully.
...
[ 3854.990122] H systemd-homed[41]: homedsshtest: changing state deactivating → inactive
...
[ 3855.089028] H systemd[1]: Stopping user-runtime-dir@60202.service...
[ 3855.121472] H systemd[1]: run-user-60202.mount: Deactivated successfully.
[ 3855.123665] H systemd[1]: user-runtime-dir@60202.service: Deactivated successfully.
[ 3855.124370] H systemd[1]: Stopped user-runtime-dir@60202.service.
[ 3855.125248] H sshd[730]: debug1: userauth-request for user homedsshtest service ssh-connection method none [preauth]
[ 3855.125248] H sshd[730]: debug1: attempt 0 failures 0 [preauth]
[ 3855.126691] H systemd[1]: Removed slice user-60202.slice.
[ 3855.128423] H systemd-logind[42]: Removed session 159.
...
[ 3855.174611] H sshd[730]: debug1: do_pam_account: called
[ 3855.202030] H systemd-homed[41]: homedsshtest: changing state inactive → activating-for-acquire
[ 3855.207883] H systemd-homework[739]: None of the supplied plaintext passwords unlock the user record's hashed passwords.
[ 3855.226523] H systemd-homed[41]: Activation failed: Password for home homedsshtest is incorrect or not sufficient for authentication.
[ 3855.227546] H systemd-homed[41]: homedsshtest: changing state activating-for-acquire → inactive
[ 3855.228360] H sshd[730]: pam_systemd_home(sshd:account): Home for user homedsshtest acquired in incomplete mode, requires later activation.
[ 3855.228577] H sshd[730]: Accepted publickey for homedsshtest from 127.0.0.1 port 45562 ssh2: ECDSA SHA256:wPPqPc0e3VA+q8dAXFR8i36yDlXGlvtfrZF23hAH2Hw
[ 3855.228577] H sshd[730]: debug1: monitor_child_preauth: user homedsshtest authenticated by privileged process
[ 3855.228577] H sshd[730]: debug1: auth_activate_options: setting new authentication options [preauth]
[ 3855.228822] H systemd-homed[41]: Got notification that all sessions of user homedsshtest ended, deactivating automatically.
[ 3855.228919] H systemd-homed[41]: Home homedsshtest already deactivated, no automatic deactivation needed.
...
[ 3855.240292] H systemd[1]: Created slice user-60202.slice.
[ 3855.242923] H systemd[1]: Starting user-runtime-dir@60202.service...
[ 3855.248137] H systemd-logind[42]: New session c2 of user homedsshtest.
[ 3855.286823] H systemd[1]: Finished user-runtime-dir@60202.service.
[ 3855.308864] H systemd[1]: Started session-c2.scope.
...
[ 3855.377452] H sshd[742]: debug1: Received SIGCHLD.
[ 3855.377544] H sshd[742]: debug1: session_by_pid: pid 743
[ 3855.377544] H sshd[742]: debug1: session_exit_message: session 0 channel 0 pid 743 exit 1
[ 3855.377544] H sshd[742]: debug1: session_exit_message: release channel 0
[ 3855.377715] H sshd[730]: debug1: session_by_tty: session 0 tty /dev/pts/1
[ 3855.377715] H sshd[730]: debug1: session_pty_cleanup2: session 0 release /dev/pts/1
[ 3855.377833] H testsuite-46.sh[727]: Called without reference on home taken, can't operate.
[ 3855.377833] H testsuite-46.sh[727]: Connection to localhost closed.
```

Full journal: https://mrc0mmand.fedorapeople.org/journals/TEST-46-PAM-fail.journal (where `-b 0` is the failed run and `-b -1` is the successful one)

/cc @poettering

**Comment by mrc0mmand** [comment]
A couple more interesting points - this is not isolated to Ubuntu/Debian, it shows in Arch Linux CI as well. Also, it's not something caused by our test image, since I can reproduce this on a "normal" Arch Linux VM with this reproducer:

```bash
# cat repro.sh 
#!/bin/bash
# vi: ts=4 sw=4 et:
set -eux
set -o pipefail

cat >/etc/ssh/sshd_config.d/20-systemd-userdb.conf <<EOF
AuthorizedKeysCommand /usr/bin/userdbctl ssh-authorized-keys %u
AuthorizedKeysCommandUser root
UsePAM yes
AcceptEnv PASSWORD
LogLevel DEBUG3
EOF
    cat >/run/systemd/system/mysshserver.socket <<EOF
[Socket]
ListenStream=4711
Accept=yes
EOF

    cat >/run/systemd/system/mysshserver@.service <<EOF
[Service]
ExecStart=-/usr/sbin/sshd -i -d -e
StandardInput=socket
StandardOutput=socket
StandardError=journal
EOF
systemctl daemon-reload
systemctl restart mysshserver.socket

if ! homectl inspect homesshtest; then
        rm -f /tmp/homed.id_ecdsa
        ssh-keygen -N '' -C '' -t ecdsa -f /tmp/homed.id_ecdsa
        NEWPASSWORD=hunter4711 homectl create \
                           --disk-size=min \
                           --luks-discard=yes \
                           --luks-pbkdf-type=pbkdf2 \
                           --luks-pbkdf-time-cost=1ms \
                           --enforce-password-policy=no \
                           --ssh-authorized-keys=@/tmp/homed.id_ecdsa.pub \
                           --stop-delay=0 \
                           homedsshtest
fi

for i in {0..100}; do
        ssh -t -t -4 -p 4711 -i /tmp/homed.id_ecdsa \
            -o "SetEnv PASSWORD=hunter4711" -o "StrictHostKeyChecking no" \
            homedsshtest@localhost true
        sleep .$RANDOM
done
```

```
# systemctl --version
systemd 256~devel (256~devel-gca90449)
# ./repro.sh
...
+ sleep .5475
+ for i in {0..100}
+ ssh -t -t -4 -p 4711 -i /tmp/homed.id_ecdsa -o 'SetEnv PASSWORD=hunter4711' -o 'StrictHostKeyChecking no' homedsshtest@localhost true
Environment:
  PASSWORD=hunter4711
  USER=homedsshtest
  LOGNAME=homedsshtest
  HOME=/
  PATH=/usr/local/sbin:/usr/local/bin:/usr/bin
  SHELL=/usr/bin/systemd-home-fallback-shell
  TERM=xterm-256color
  XDG_SESSION_INCOMPLETE=1
  SYSTEMD_HOME=1
  SYSTEMD_HOME_SUSPEND=0
  MOTD_SHOWN=pam
  MAIL=/var/spool/mail/homedsshtest
  XDG_SESSION_ID=67
  XDG_RUNTIME_DIR=/run/user/60202
  XDG_SESSION_TYPE=tty
  XDG_SESSION_CLASS=user-incomplete
  SSH_CLIENT=127.0.0.1 44358 4711
  SSH_CONNECTION=127.0.0.1 44358 127.0.0.1 4711
  SSH_TTY=/dev/pts/2
Called without reference on home taken, can't operate.
Connection to localhost closed.
```

**Comment by mbiebl** [comment]
fwiw, I need to take back my comment earlier https://github.com/systemd/systemd/issues/31518#issuecomment-1973022887
Reverting that PR does not make the test pass reliably, but just more likely.
Where before it failed basically a 100%, with the PR reverted the rate is about 50%.

This supports @mrc0mmand suspicion, that this is a race somewhere.

**Comment by mbiebl** [comment]
@poettering let's continue the discussion here, after #31544 has been merged.

This affects CI (a good chunk of ubuntu-ci failures is because of upstream/TEST-46-HOMED), which is bad as people start ignoring failures.

I would like to see this treated with a higher priority to unbreak CI.

**Comment by AdrianVovk** [comment]
Looking into this. Looks like the problem is definitely the open FD (the thing acquired by `acquire_home` in `pam_systemd_home`) gets lost/closed somehow before `pam_sm_open_session` (or at least homed _thinks_ that's what happened) and so it drops all references on the home area, which ultimately leads to the error in fallback-shell

There's a couple of places where this can happen:
- The session is closed (`pam_sm_close_session` -> `release_home_fd` -> overwrite `systemd-home-fd-<username>` -> close fd)
    - We know this isn't happening at the time of the bug because we log when we're closing the session
- The entire PAM handle is destroyed
   - I checked in openssh's implementation, this isn't being called in this case
- The `systemd-home-fd-<username>` data entry on the PAM handle is overwritten
   - It's only ever written from `release_home_fd` (only called from `close_session`, so not it) and `acquire_home` (which is a no-op of `systemd-home-fd-<username>` is already set to something)

**Comment by AdrianVovk** [comment]
As far as I can tell there's really no spot where we could be dropping the FD, ESPECIALLY in a racy manner.

So here's a new idea. If you look at the log, we actually fail to call into ReleaseHome() for some reason. We get an access denied error (why = unknown so far. Need to investigate)

So a theory: a previous instance of the PAM module obtains an FD, does whatever it needs to do, then drops the FD and tries to synchronize on everything getting deactivated. But we fail to synchronize! So now homed is busy. Meanwhile, the test continues, starts a new instance of sshd and PAM, which fills homed's sd_event queue with dbus messages and other events it needs to respond to. Note that handling the closed FD is an idle priority event, so the events sshd ends up queuing up by trying to talk to homed are higher priority

Ultimately, the ref_eof event is starved for so long that it ends up being executed only after the next instance of sshd's PAM module tries to activate the home again. And here's the race condition: if the ref_eof event gets lucky and is handled before the RefUnrestricted, everything works fine. The reference is dropped, and then immediately picked back up. However if the ref_eof event gets really unlucky, it gets handled _after_ RefUnrestricted!

If ref_eof is handled after RefUnrestricted: RefUnrestricted tries to create the appropriate FIFO event source, but skips that step since one already exists. It then opens the writing end and returns it via dbus. Then ref_eof gets to execute, and all it does is wipe out the FIFO event source and declare the home no-longer-referenced (despite the fact that it is referenced!). Thus, the file descriptor returned by RefUnrestricted is doing nothing.

**Comment by AdrianVovk** [comment]
OK I think I have confirmation of my theory. Here's a reproducer of the race condition I describe above. It seems to race 100% of the time

```python3
import os
from gi.repository import Gio, GLib
bus = Gio.bus_get_sync(Gio.BusType.SYSTEM)

def ref(): # Take ref on home area; drop w/ os.close
    return bus.call_with_unix_fd_list_sync("org.freedesktop.home1", "/org/freedesktop/home1/home/homedsshtest", "org.freedesktop.home1.Home", "RefUnrestricted", GLib.Variant("(b)", [True]), None, 0, -1).out_fd_list.get(0)

def work(): # Queues up a bunch of requests to simulate "work"
    for i in range(0, 100):
        bus.call("org.freedesktop.home1", "/org/freedesktop/home1/home/homedsshtest", "org.freedesktop.home1.Home", "Authenticate", GLib.Variant("(s)", ["{}"]), None, 0, -1)

def sync_has_ref(): # Let homed catch up on events, then return True/False depending on if homed thinks we're holding a ref or not
    # Note: Release() will always be processed *after* the ref_eof event, because it has a lower priority. Thus we can use it as a sync point
    try:
        bus.call_sync("org.freedesktop.home1", "/org/freedesktop/home1/home/homedsshtest", "org.freedesktop.home1.Home", "Release", None, None, 0, -1)
        return False
    except:
        return True

def reproduce():
    # start first instance of sshd
    work()       # Trying to authenticate, failing
    fd = ref()   # Take a reference on the home directory. Now an "incomplete session" 
    work()       # fallback-shell trying to authenticate, maybe succeeding, exiting, etc
    os.close(fd) # We're done w/ the home area. Time to clean up
    # WHOOPS! We fail to call Release() and sync w/ homed for some reason
    # sshd now exits
    
    # If we successfully raced, at this point we've released the ref but homed hasn't noticed
    # yet! It has yet to handle that event.
    
    # start second instance of sshd. Same as above
    work()
    fd = ref()
    
    # If we successfully raced, at this point we've re-acquired the ref (homed already processed
    # this fact since it gave us the FD already) but homed _still_ hasn't processed the earlier
    # event releasing the old ref. Let's now let homed catch up and check if homed thinks the
    # home is ref'd
    assert sync_has_ref(), "REPRODUCED! We're holding the FD, but homed thinks we're not!"
    
    # (rest of second sshd session)
    work()
    os.close(fd)

while True:
    reproduce()
    assert not sync_has_ref() # Sync up for next attempt
    print("No race")
```

So I think we can take two distinct approaches here.

1. Make it impossible for homed to race in this way, by somehow ensuring that the fd close event is handled early and often. Making it higher-than-normal priority rather than super-idle is a fix. I need input from @poettering
2. Figure out why `Release()` is failing and fix it

I'll pursue approach 2 for now and that should fix the CI failures. Once that's done, however, I'd open a bug report about the fact that homed's world-view can drift out of consistency so we can discuss solutions for that somewhat distinct issue

**Comment by AdrianVovk** [comment]
Investigated further. We're getting permission denied errors in the failure case because `sd-pam` (PID 717 in the failing case) drops privileges. Supposedly these privileges aren't necessary to close the session (according to a comment in src/core/exec-invoke.c), but I think that's blatantly incorrect... `ReleaseHome()` is callable by anyone at the dbus level, but homed itself will only allow privileged users (i.e. uid 0) to call it. Thus, we see no complaints from dbus-daemon, but homed returns an access denied error. `ReleaseSession()` is denylisted at the dbus level to anyone but root. So logind never even gets the request: dbus-daemon immediately denies it, loudly complains about it, and returns an access denied error. In both cases, we call these from `pam_sm_close_session` and as an unprivilged user so the access denied checks are expected and should be happening 100% of the time - the fact that it's _not_ always failing is a bug

My theory of what's happening: (after the revert?) we're leaving an sd-bus connection open across the fork. Somehow this makes it possible to (racily) trick dbus's credentials checking into thinking we're still running as root rather than our unprivileged user. So sometimes dbus gets confused, decides we're still privileged, and lets us call the functions (this is the "success" case. call the functions -> sync w/ homed -> no race condition with the reference fd -> test passes). Other times, dbus doesn't get confused, knows that we're actually unprivileged, and doesn't let us call the functions (this is the failure case: no functions called -> no syncing -> race w/ reference fd -> fail the test)

CCing @mrc0mmand and @yuwata for help here, not sure what the best way to fix the sd-bus-stuff-being-passed-into-pam logic, nor am I really understanding the prior back and forth (PR, then reverts) that has already happened. So, help would be appreciated

We probably shuoldn't try to drop privileges when running `(sd-pam)` stuff? I'm also really doubting that it's safe to fork half-way through a PAM conversation; why not fork first, then run all the pam things inside of the child? Not sure who to CC for this

**Comment by yuwata** [comment]
Not sure it is related, but there is a proposal for PAM session, #31033.
